File: data_structures_and_algorithms/Code_401/class-03/solutions/python/array-binary-search/array_binary_search.py
def binary_search_i(nums, key):

    left = 0

    right = len(nums) - 1

    while left <= right:

        mid = (right - left) // 2 + left

        candidate = nums[mid]

        if candidate == key:
            return mid
        elif key < candidate:
            right = mid - 1
        else:
            left = mid + 1

    return -1


def binary_search(nums, key, left=0, right=None):

    if right is None:
        right = len(nums) - 1

    if right < left:
        return -1

    mid = (right - left) // 2 + left

    candidate = nums[mid]

    if candidate == key:

        return mid

    elif key < candidate:

        return binary_search(nums, key, left, mid - 1)

    else:

        return binary_search(nums, key, mid + 1, right)

File: c-sharp/CodeChallenge3/ArrayBinarySearch.cs
namespace CodeChallenges
{
  public class CodeChallenge3
  {
    public static int BinarySearch(int[] array, int key)
    {
      int left = 0;
      int right = array.Length - 1;

      while (left <= right)
      {
        int mid = left + (right - left) / 2;

        if (array[mid] == key)
        {
          while (mid > 0 && array[mid - 1] == key)
          {
            mid--;
          }
          return mid;
        }

        if (array[mid] < key)
        {
          left = mid + 1;
        }
        else
        {
          right = mid - 1;
        }
      }

      return -1;
    }
  }
}
